### ISSUE TITLE
[Feature Request] LDAP for Teams

### ISSUE THREAD

**Issue by Zegorax** [padloc/padloc#256]
Would it be possible to have an LDAP / Active Directory integration for Teams ?

Some ideas as a brainstorm :
- Login with LDAP username / passwords
- Filters for login
- Attribute mapping when a custom LDAP server is used.
- Group synchronization with tags (RO / RW)
- Group synchronization with vaults (RO / RW)
- Group synchronization with organization admin role

This approach would be compatible with RBAC as well.

**Comment by MaKleSoft** [comment]
I do think we can add some value with some form of AD integration, although how far we can go is limited somewhat by the very security-centric design of Padloc. One thing that can definitely not be "synced" in any way is the master password, and onboarding of new users will likely require at least some oversight by a system admin.

Can you perhaps lay out how you would envision using Padloc with an AD/LDAP system? What kind of features would you expect, and what degree of automation? (You already wrote down some bullet points, but more details would be nice). Feedback from others appreciated as well!

**Comment by xtonousou** [comment]
* Login with LDAP / AD username and password is the most crucial
* Group for Administrators and Users
* Administrators must not be able to look at others' passwords (Administrators or Users), but can manage and configure Padloc

**Comment by MaKleSoft** [comment]
> Login with LDAP / AD username and password is the most crucial

This one is probably the most problematic, since making sure that every users master password is both strong and **truly** secret is paramount. Simply reusing the AD password would be a huge downgrade in security. I'm also not sure how the password would be synced with Padloc since I imagine/hope no AD implementation would store it in plain text? I imagine the other direction would be more sensible, i.e. users choose their master password when signing up for Padloc (as part of the provisioning process) and then store their AD password somewhere within Padloc. I don't know enough about AD architecture to know if this is feasible though

> Administrators must not be able to look at others' passwords (Administrators or Users), but can manage and configure Padloc

This is already the case in the current version. Especially the first part is something that (although somewhat tricky to pull off) was a strict requirement for the Padloc 3 design from the start.

**Comment by Elhitch** [comment]
@MaKleSoft I could argue with your statement that reusing the AD/LDAP password would be a huge downgrade in security.

People and corporations use password managers to make password management easier, not harder; having a separate password for both services means that people need to remember more which means they are more likely to use weak passwords for both. Additionally, storing AD/LDAP passwords in the password manager lacks any purpose at all since normally employees need to be authenticated to the OS / VPN / etc. in order to access corporate resources which requires AD/LDAP password in the first place. Why would they need to store it in their password manager?

Additionally, for organizations that require strict security, there are password requirement policies which are pretty well developed and any system administrator know how to configure. The Kerberos protocol is well developed and can be used for integration here.

Not to mention the benefit of AD/LDAP integration leading to better monitoring of the entire environment since SIEMs can understand AD/LDAP logs and security analysts are much better trained in spotting anomalies in AD/LDAP logs. Does Padloc even have an authentication log which can be sent to third-party monitoring solutions?

I see no excuse from the security point of view not to implement this. Instead, you can implement 2FA/MFA in order to make sure that authenticated entities are actually authorized.

**Comment by MaKleSoft** [comment]
> People and corporations use password managers to make password management easier, not harder; having a separate password for both services means that people need to remember more which means they are more likely to use weak passwords for both. Additionally, storing AD/LDAP passwords in the password manager lacks any purpose at all since normally employees need to be authenticated to the OS / VPN / etc. in order to access corporate resources which requires AD/LDAP password in the first place. Why would they need to store it in their password manager?

I fully agree! Finding a good balance between security and usability/convenience is extremely important in any real-world scenario, especially if non-technical users are involved. But at the same time, it's important to not let people stay complacent but try to educate them and guide them towards better practices instead (by "people" I mean not just end-users, but sysadmins and corporate decision makers as well) . Half-assed security is just as bad (or even worse) than no security at all.

Do you want the key to your garden shed to unlock your safe-deposit box as well? Different things require different levels of security, and a password manager potentially holds an organisation's most sensitive information by far.

> Additionally, for organizations that require strict security, there are password requirement policies which are pretty well developed and any system administrator know how to configure. The Kerberos protocol is well developed and can be used for integration here.

Unfortunately based on my own experience the average corporate password policy is not nearly at the level that should be required for a password manager's master password. The difference between AD/LDAP passwords and the PM master password is that the latter is not only used for authentication, but for **encryption** as well, and therefore should be able to withstand a brute-force attack against a local copy of the encrypted data. Kerberos (or any other authentication scheme) doesn't really help us here, I'm afraid.

> Does Padloc even have an authentication log which can be sent to third-party monitoring solutions?

Not yet, but it's on our roadmap.

> I see no excuse from the security point of view not to implement this. Instead, you can implement 2FA/MFA in order to make sure that authenticated entities are actually authorised.

Again, authentication is only part of the problem. The fact is that **it is impossible to synchronise passwords between systems without violating the zero-knowledge principle**. I can go into more technical detail if you want, but trust me when I say that we've spent quite a few hours wrecking our brains about this 🤓.

Now, having said all this, we are aware that syncing passwords is a hard requirement for a lot of businesses, and **even though we will always recommend against it, we do plan to add support for it eventually**, at least for enterprise/on-premise users who are willing to make the required compromises (effectively foregoing the benefits of zero-knowledge). So stay tuned 😉
